### PR TITLE
Add content presenters fluent textbox

### PR DESCRIPTION
**Pull request by jmacato**
## What does the pull request do?
Implements a content hosting for the left and right sides of the fluent TextBox.. Useful for applications like password hide button, etc etc.

### CODE REVIEW

**Comment by maxkatz6** [comment]
I see it would be really useful to improve [NumericUpDown](https://github.com/AvaloniaUI/Avalonia/issues/4039#issuecomment-644134993) and [AutoCompleteBox](https://github.com/AvaloniaUI/Avalonia/issues/4023) controls. Thanks!
But I am thinking about default behavior for TextBox - shouldn't it contain ["clear"](https://github.com/AvaloniaUI/Avalonia/issues/4022) and "show password" (in passwordbox mode) buttons out of box?
If yes, should it be placed as "InnerRightContent"?

**Comment by jmacato** [comment]
@maxkatz6 yep, @danwalmsley suggested the same :D

File: samples/ControlCatalog/Pages/TextBoxPage.xaml [hunk 1]
```diff
@@ -1,58 +1,74 @@
...
-                 Watermark="Floating Watermark"
-                 UseFloatingWatermark="True"
-                 Text="Lorem ipsum dolor sit amet, consectetur adipiscing elit."/>
-
-        <TextBox Width="200"
-                 Watermark="Password Box"
-                 UseFloatingWatermark="True"
-                 PasswordChar="*"
-                 Text="Password" />
+        <TextBox Width="200" Watermark="Floating Watermark" UseFloatingWatermark="True" Text="Lorem ipsum dolor sit amet, consectetur adipiscing elit."/>
```
Comment (danwalmsley) [inline samples/ControlCatalog/Pages/TextBoxPage.xaml:37]: We need to remove unrelated formatting changes.

File: src/Avalonia.Controls/TextBox.cs [hunk 2]
```diff
@@ -123,6 +132,10 @@ public UndoRedoState(string text, int caretPosition)
         static TextBox()
         {
             FocusableProperty.OverrideDefaultValue(typeof(TextBox), true);
+            ShowPasswordProperty.Changed.Subscribe((x)=>
```
Comment (danwalmsley) [inline src/Avalonia.Controls/TextBox.cs:135]: noting happening here?

File: src/Avalonia.Controls/TextBox.cs [hunk 3]
```diff
@@ -285,7 +298,7 @@ public string SelectedText
                 else
                 {
                     HandleTextInput(value);
-                } 
+                }
```
Comment (danwalmsley) [inline src/Avalonia.Controls/TextBox.cs:301]: unrelated change

File: src/Avalonia.Controls/TextBox.cs [hunk 4]
```diff
@@ -326,6 +339,28 @@ public bool UseFloatingWatermark
...
+            get { return GetValue(InnerRightContentProperty); }
+            set { SetValue(InnerRightContentProperty, value); }
+        }
+
+        public bool ShowPassword
+        {
+            get { return GetValue(ShowPasswordProperty); }
+            set
+            {
+
```
Comment (danwalmsley) [inline src/Avalonia.Controls/TextBox.cs:359]: whitespace...

use => for get / set.

File: src/Avalonia.Controls/TextBox.cs [hunk 5]
```diff
@@ -756,7 +791,7 @@ protected override void OnPointerReleased(PointerReleasedEventArgs e)
                     // if it did not, we change the selection to where the user clicked
                     var firstSelection = Math.Min(SelectionStart, SelectionEnd);
                     var lastSelection = Math.Max(SelectionStart, SelectionEnd);
-                    var didClickInSelection = SelectionStart != SelectionEnd && 
+                    var didClickInSelection = SelectionStart != SelectionEnd &&
```
Comment (danwalmsley) [inline src/Avalonia.Controls/TextBox.cs:794]: ?

**Review by danwalmsley — CHANGES_REQUESTED** [review]
A huge amount of unrelated formatting changes makes it very hard to see the wood for the trees.

The hide/show password needs to be integrated to a button or style that applies that button.

Follow the fluent icons style too.

**Review by danwalmsley — CHANGES_REQUESTED** [review]

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 6]
```diff
@@ -75,7 +78,7 @@ public class TextPresenter : Control
         static TextPresenter()
         {
             AffectsRender<TextPresenter>(SelectionBrushProperty);
-            AffectsMeasure<TextPresenter>(TextProperty, PasswordCharProperty, 
+            AffectsMeasure<TextPresenter>(TextProperty, PasswordCharProperty, ShowPasswordTextProperty,
```
Comment (danwalmsley) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs:81]: Why affects measure? id have thought affectsrender?

File: src/Avalonia.Controls/TextBox.cs [hunk 7]
```diff
@@ -95,6 +95,15 @@ public class TextBox : TemplatedControl, UndoRedoHelper<TextBox.UndoRedoState>.I
             AvaloniaProperty.RegisterDirect<TextBox, string>(nameof(NewLine),
                 textbox => textbox.NewLine, (textbox, newline) => textbox.NewLine = newline);
 
+        public static readonly StyledProperty<object> InnerLeftContentProperty =
+            AvaloniaProperty.Register<TextBox, object>(nameof(InnerLeftContent));
+
+        public static readonly StyledProperty<object> InnerRightContentProperty =
+            AvaloniaProperty.Register<TextBox, object>(nameof(InnerRightContent));
+
+        public static readonly StyledProperty<bool> ShowPasswordTextProperty =
```
Comment (danwalmsley) [inline src/Avalonia.Controls/TextBox.cs:104]: Rename to RevealPassword?

File: src/Avalonia.Controls/TextBox.cs (hunk 5 above)
Comment (danwalmsley) [reply src/Avalonia.Controls/TextBox.cs:786]: unrelated change

File: src/Avalonia.Themes.Fluent/TextBox.xaml [hunk 8]
```diff
@@ -1,6 +1,14 @@
 <Styles xmlns="https://github.com/avaloniaui" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
   <Styles.Resources>
     <Thickness x:Key="TextBoxTopHeaderMargin">0,0,0,4</Thickness>
+    <DrawingGroup x:Key="eye_show_regular">
```
Comment (danwalmsley) [inline src/Avalonia.Themes.Fluent/TextBox.xaml:13]: iv not seen any other resources named with this style.

File: src/Avalonia.Themes.Fluent/TextBox.xaml [hunk 9]
```diff
@@ -147,4 +159,33 @@
...
+          <Panel Width="16" Height="16">
+            <DrawingPresenter Drawing="{DynamicResource eye_show_regular}" IsVisible="{Binding $parent[TextBox].ShowPasswordText}"/>
+            <DrawingPresenter Drawing="{DynamicResource eye_hide_regular}" IsVisible="{Binding !$parent[TextBox].ShowPasswordText}"/>
+          </Panel>
+        </ToggleButton>
+      </Template>
+    </Setter>
+  </Style>
+
+  <Style Selector="TextBox.leftClearButton">
```
Comment (danwalmsley) [inline src/Avalonia.Themes.Fluent/TextBox.xaml:182]: imo i think we only need the right hand one.

**Review by danwalmsley — CHANGES_REQUESTED** [review]

**Review by danwalmsley — APPROVED** [review]
LGTM

Styling of buttons will be improved by @maxkatz6

**Review by danwalmsley — CHANGES_REQUESTED** [review]
Please add support for default template.

**Comment by Gillibald** [comment]
Any reason why there is no PasswordBox that covers all this? The current implementation isn't secure and I don't see why a TextPresenter should have functionality specific to a PasswordBox. I don't think we can support those PasswordBox related properties forever. Just my two cents.

**Comment by jmacato** [comment]
@Gillibald we should do it in another PR.. but right now there isnt any secure memory handling for passwords in dotnet core.. SecureString is a huge farce and deprecated... One way to do it is to have a special memory zero'ing logic and pinning to prevent GC from copying it somewhere or whatever.

**Comment by danwalmsley** [comment]
> @Gillibald Right now there isnt any secure memory handling for passwords in dotnet core.. SecureString is a huge farce and deprecated... One way to do it is to have a special memory zero'ing logic and pinning to prevent GC from copying it somewhere or whatever.
> 
> But i agree with textpresenter not doing this password logic and we could address that in another PR

We had that discussion about passwordbox a long time ago. It was decided to just do this simple password char thing with textbox, which IIRC matches winforms.

File: src/Avalonia.Themes.Fluent/TextBox.xaml [hunk 10]
```diff
@@ -46,57 +54,60 @@
...
-                    </Panel>
-                  </ScrollViewer>
-                </DataValidationErrors>
-              </DockPanel>
+              <Grid ColumnDefinitions="Auto,*,Auto" >
+                <ContentPresenter Grid.Column="0" Grid.ColumnSpan="1" Content="{TemplateBinding InnerLeftContent}"/>
+                <DockPanel Grid.Column="1" Grid.ColumnSpan="1" Margin="{TemplateBinding Padding}">
+                  <TextBlock Name="floatingWatermark"
+                            Foreground="{DynamicResource SystemAccentColor}"
+                            FontSize="{DynamicResource FontSizeSmall}"
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/TextBox.xaml:63]: FontSizeSmall is missed in Fluent theme. You need to replace it with ControlContentThemeFontSize (there is not "small" equivalent) or restore "TemplateBinding FontSize" as it was before that PR. 
```suggestion
                            FontSize="{TemplateBinding FontSize}"
```

**Review by maxkatz6 — CHANGES_REQUESTED** [review]

File: src/Avalonia.Themes.Fluent/TextBox.xaml [hunk 11]
```diff
@@ -156,4 +166,25 @@
     <Setter Property="Cursor" Value="IBeam" />
```
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/TextBox.xaml:166]: Cursor=IBeam should be set only on inner DockPanel (there are two of them in the template), so InnerLeft and InnerRight will have default cursor.

**Review by maxkatz6 — CHANGES_REQUESTED** [review]

File: src/Avalonia.Themes.Fluent/TextBox.xaml (hunk 10 above)
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/TextBox.xaml:63]: Fixed in https://github.com/AvaloniaUI/Avalonia/pull/4487

File: src/Avalonia.Themes.Fluent/TextBox.xaml (hunk 11 above)
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/TextBox.xaml:166]: Fixed in https://github.com/AvaloniaUI/Avalonia/pull/4487

**Review by danwalmsley — APPROVED** [review]
LGTM, would like an approval from @grokys  before merging... perhaps we should be registering contentpresenters.

**Pull request merged**